[RISCV][GISEL] Legalize and regbankselect vector typed G_IMPLICIT_DEF

michaelmaitland opened the pull request:
Let instruction selection do what it wants with this opcode for scalable vector types that are already legal.

michaelmaitland committed:
[RISCV][GISEL] Legalize and regbankselect vector typed G_IMPLICIT_DEF

michaelmaitland committed:
[RISCV][GISEL][NFC] Add typeIsLegalBoolVec and typeIsLegalIntOrFPVec helpers

tschuett commented:
Did you try legalization of illegal types, e.g.,  LLT::scalable_vector(30, s1)?

tschuett commented:
Disallow illegal types! But the legalizer should transform illegal types into legal types. For scalable vectors there is not yet good test coverage.

michaelmaitland commented:
> Disallow illegal types! But the legalizer should transform illegal types into legal types. For scalable vectors there is not yet good test coverage.

I have added a TODO in the legalizer to make illegal types of G_IMPLICIT_DEF into legal types, and I have updated this PR to specify that this patch only legalizes vector types that are considered legal.

michaelmaitland commented:
ping

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVRegisterBankInfo.cpp:
```diff
@@ -351,6 +351,11 @@ RISCVRegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
     if (anyUseOnlyUseFP(Dst, MRI, TRI))
       Mapping = getFPValueMapping(MRI.getType(Dst).getSizeInBits());
+
+    LLT DstTy = MRI.getType(Dst);
```
Move the declaration of `LLT DstTy` earlier and use it on line 353 too.

michaelmaitland commented:
ping, all my open GISEL PRs depend on this one :)

topperc reviewed (approved):
LGTM

michaelmaitland commented:
Committed in 33cf27b0a3f307f4c57e868fedb72f4d65ba72d9 and ea798a79000f5df53b469ecb491d7d21479d6f24

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:
```diff
@@ -40,6 +40,30 @@ static LegalityPredicate typeIsScalarFPArith(unsigned TypeIdx,
+  LegalityPredicate HasV = [=, &ST](const LegalityQuery &Query) {
+    return ST.hasVInstructions();
+  };
+  return all(typeInSet(TypeIdx, BoolVecTys), HasV);
```
I missed it in the review, but this part of regular vectors applies to bools too

```
           (Query.Types[TypeIdx].getElementCount().getKnownMinValue() != 1 ||    
            ST.getELen() == 64); 
```

michaelmaitland replied:
I will add it

michaelmaitland replied:
3305f5a980363d7e224094bc42bcbebcf162618e
